OVS-4358 - GeoScale

kvanhijf opened the pull request:
Initial commit to create branch
Removed possibility to run setup remote
WIP

kvanhijf committed:
OVS-4358 - GeoScale
Initial commit to create branch
Removed possibility to run setup remote

kvanhijf committed:
OVS-4358 - Initial commit with some cleanup actions
Removed possiblity to setup node remotely
Converted cfg files to json
Removed obsolete parameters
Added key, value validation for unattended installation
Fixed issue when avahi not installed and unattended installation

kvanhijf committed:
OVS-4348

kvanhijf committed:
OVS-4348 - Fixed issues mentioned in code review

kvanhijf committed:
OVS-4358 - Implementation for predefined rabbitmq and arakoon clusters

kvanhijf committed:
Merge branch 'master' into ovs-4348-geoscale

Commit:
OVS-4348-geoscale automation

Commit:
Merge remote-tracking branch 'origin/ovs-4348-geoscale' into ovs-4348-geoscale

khenderick commented on ovs/extensions/db/arakoon/ArakoonInstaller.py:
```diff
@@ -414,7 +437,11 @@ def _get_free_ports(client):
-        ports = System.get_free_ports(EtcdConfiguration.get('/ovs/framework/hosts/{0}/ports|arakoon'.format(node_name)), exclude_ports, 2, client)
+        key = '/ovs/framework/hosts/{0}/ports|arakoon'.format(node_name)
+        if EtcdConfiguration.exists(key):
+            ports = System.get_free_ports(EtcdConfiguration.get(), exclude_ports, 2, client)
```
You're missing key here.

khenderick commented on ovs/extensions/db/arakoon/ArakoonInstaller.py:
```diff
@@ -187,6 +190,26 @@ def __init__(self):
 
     @staticmethod
+    def get_etcd_config_path():
+        return ArakoonInstaller.ETCD_CONFIG_PATH
```
What's the purpose of these methods? You could just simply get or set the property itself. And if you still need a setter and getter, it might be a good idea to actually use setters and getters.

Commit:
OVS-4348-geoscale automation

kvanhijf committed:
OVS-4358 - Predefined ovsdb arakoon cluster (WIP)

kvanhijf committed:
OVS-4358 - Bugfixes in initial implementation for external ovsdb cluster

Commit:
Merge master

Commit:
Geoscale installation updates

kvanhijf committed:
Merge branch 'master' into ovs-4348-geoscale

kvanhijf committed:
OVS-4358 - Fixed merge issue

kvanhijf committed:
OVS-4358 - Introduction of arakoon cluster type ENUM

kvanhijf committed:
OVS-4358 - Voldrv arakoon clusters
Only tested internally managed voldrv cluster

Commit:
Geoscale updates

khenderick commented on ovs/extensions/db/arakoon/ArakoonInstaller.py:
```diff
@@ -719,3 +722,64 @@ def restart_cluster_remove(cluster_name, remaining_ips):
+            if not EtcdConfiguration.exists(key):
+                raise RuntimeError('No cluster metadata found for {0}'.format(cluster_name))
+            values = json.loads(EtcdConfiguration.get(key, raw=True))
+            if values['in_use'] == True:
```
you could use `EtcdConfiguration.get(key)` that would do the `json.loads(...)` etc for you. Even more, you can also use stuff like `EtcdConfiguration.get('/ovs/foo/bar|some.key')` which will even traverse the json object and return the value of that key. The set works exactly the same. More documentation and examples are in the EtcdConfiguration class' docstring.

kvanhijf committed:
OVS-4358 - Implementation of new claim_cluster and is_internal calls of ArakoonInstaller class

kvanhijf committed:
OVS-4358 - Initial implementation of service.is_internal dynamic property
Lots of TODOs added and still lots of work to be done

Commit:
Geoscale updates

kvanhijf committed:
OVS-4358 - Introduction of enumerator for servicetypes
Removed ArakoonInstaller.is_internal because was based on cluster_type instead of cluster name
Removed create_or_extend_cluster in albacontroller and separated code again where it was called originally
Introduced both ArakoonClusterMetadata claim and unclaim methods
Added lots more docstrings
Merged nsm_checkup calls again

kvanhijf committed:
OVS-4358 - ArakoonInstaller.create_cluster no longer claims the created cluster

kvanhijf committed:
OVS-4358 - 2 bugfixes

kvanhijf committed:
Merge branch 'master' into ovs-4348-geoscale

kvanhijf committed:
OVS-4358 - Fixes during merge

kvanhijf committed:
OVS-4358 - Fixed issues regarding removal of backend using internally managed arakoon clusters
Fixed issue regarding removal of vpool when corresponding backend is no longer available

kvanhijf committed:
OVS-4360 - Reworked some occurrences of storagerouter is not None to is_internal is True
Fixed issue when demoting a node with an NSM cluster present

kvanhijf committed:
OVS-4359 - Parameter validation for unattended installation
Also possible to force master or extra using unattended installation
Also possible to setup cluster with different passwords for different nodes

kvanhijf committed:
OVS-4359 - Bugfix

kvanhijf committed:
OVS-4451 - Bugfix of the bugfix

kvanhijf committed:
OVS-4359 - Fix for unattended install for KVM

kvanhijf committed:
Merge branch 'master' into ovs-4348-geoscale

kvanhijf committed:
OVS-4451 - Adjustments required for hostname changes (OVS-4419)
Necessary after merge conflicts

kvanhijf committed:
OVS-4451 - Bugfix after merge conflict

khenderick commented on ovs/extensions/db/arakoon/ArakoonInstaller.py:
```diff
@@ -244,19 +319,44 @@ def clean_leftover_arakoon_data(ip, directories):
 
-        client = SSHClient(ip)
+        EtcdConfiguration.set('/ovs/framework/stores', {'persistent': 'pyrakoon',
+                                                        'volatile': 'memcache'})
```
It doesn't make sense to overwrite the volatile configuration when you set up a new cluster.

khenderick replied:
Will this be executed even if the customer wants to use external clusters set up by some whatever script using whatever other code?

The issue is that this piece is not related to arakoon at all. If the issue is that these keys are required for the arakooninstaller extension to work, why not just execute the etcd installer's initialize method before start using the arakooninstaller?

khenderick commented on ovs/extensions/db/arakoon/ArakoonInstaller.py:
```diff
@@ -390,14 +517,68 @@ def deploy_cluster(cluster_name, node_ip):
+        cluster_type = cluster_type.upper()
+        if cluster_type not in ServiceType.ARAKOON_CLUSTER_TYPES:
+            raise ValueError('Unsupported arakoon cluster type provided. Please choose from {0}'.format(', '.join(ServiceType.ARAKOON_CLUSTER_TYPES)))
+        if not EtcdConfiguration.exists('/ovs/arakoon', raw=True):
```
Maybe `dir_exists` is more appropriate here.

khenderick commented on ovs/extensions/db/arakoon/ArakoonInstaller.py:
```diff
@@ -414,7 +595,11 @@ def _get_free_ports(client):
                 except:
                     logger.error('  Could not load port information of cluster {0}'.format(cluster_name))
-        ports = System.get_free_ports(EtcdConfiguration.get('/ovs/framework/hosts/{0}/ports|arakoon'.format(node_name)), exclude_ports, 2, client)
+        key = '/ovs/framework/hosts/{0}/ports|arakoon'.format(node_name)
```
This change might become obsolete if you call the etcdconfiguration's initialize before you start using the arakooninstaller. Then the `ARAKOON_START_PORT` isn't needed either. And you have the benefit that you might be able to adapt the initialize method to overwrite arakoon ports before, so you can setup clusters on other ports.

khenderick commented on ovs/extensions/storage/persistentfactory.py:
```diff
@@ -30,16 +30,25 @@ class PersistentFactory(object):
+                metadata_key = '/ovs/arakoon/{0}/metadata'.format(cluster)
+                if EtcdConfiguration.exists(metadata_key):
+                    metadata = EtcdConfiguration.get(metadata_key)
+                    if metadata['type'].upper() == 'FWK':
```
Use ArakoonMetadata object and use the enumerators. I'd place this in the correct if below as well.

khenderick commented on ovs/extensions/storageserver/storagedriver.py:
```diff
@@ -176,6 +176,9 @@ def load(service):
         """
 
+        if service.storagerouter is None:
+            return None
```
It's better to raise an error here. This only should happen when we made a logical error in our code so we should see that error fairly soon. If we return None, it might go unnoticed.

khenderick commented on ovs/lib/mdsservice.py:
```diff
@@ -636,7 +668,7 @@ def get_mds_storagedriver_config_set(vpool, check_online=False):
                 except UnableToConnectException:
                     continue
-            mds_service, load = MDSServiceController.get_preferred_mds(storagerouter, vpool, include_load=True)
+            mds_service, load = MDSServiceController.get_preferred_mds(storagerouter, vpool)
```
A TypeError error might be raised here when None is returned. It might be better to either validate the return value, or always return a tuple.

khenderick commented on ovs/lib/storagedriver.py:
```diff
@@ -120,31 +129,44 @@ def on_demote(cluster_ip, master_ip, offline_node_ips=None):
             print '* Shrink StorageDriver cluster'
-            ArakoonInstaller.shrink_cluster(cluster_ip, 'voldrv', offline_node_ips)
+            clusters = ArakoonInstaller.get_arakoon_metadata_by_cluster_type(cluster_type=ServiceType.ARAKOON_CLUSTER_TYPES.SD, in_use=True)
+            if len(clusters) != 1:
```
Should check for `> 0`.

khenderick commented on ovs/lib/storagedriver.py:
```diff
@@ -165,71 +187,95 @@ def manual_voldrv_arakoon_checkup():
+                    raise RuntimeError('Could not find any Storage Router with a DB role')
+
+                storagerouter, partition = available_storagerouters.items()[0]
+                cluster_name = 'voldrv'
```
You should make sure that if another cluster would be called 'voldrv' that this code can either generate another name and continue, or that the create_cluster raises that such cluster already exists. Now it might just silently mess up configurations.

khenderick commented on ovs/lib/storagedriver.py:
```diff
@@ -165,71 +187,95 @@ def manual_voldrv_arakoon_checkup():
+
+            cluster_name = metadata[0].cluster_id
+            logger.info('Claiming {0} managed arakoon cluster: {1}'.format('externally' if storagerouter is None else 'internally', cluster_name))
+            metadata[0].claim()
```
We should have locking here. This cluster might already be claimed by another codepath.

khenderick commented on ovs/lib/storagerouter.py:
```diff
@@ -371,7 +378,7 @@ def add_vpool(parameters):
                 if preset_name not in [preset['name'] for preset in backend_info['presets']]:
                     raise RuntimeError('Given preset {0} is not available in backend {1}'.format(preset_name, backend_guid))
-                task_id = ovs_client.get('/alba/backends/{0}/get_config_metadata'.format(backend_guid))
+                task_id = ovs_client.get('/alba/backends/{0}/get_arakoon_config'.format(backend_guid))
```
I assume you have obsoleted the old call with versions?

khenderick commented on ovs/lib/tests/test_mdsservice.py:
```diff
@@ -227,7 +227,7 @@ def _create_vdisks_for_mds_service(self, amount, start_id, mds_service=None, vpo
         vdisks = {}
         storagedriver_id = None
-        if mds_service is not None:
+        if mds_service is not None and mds_service.service.storagerouter_guid is not None:
```
It should use the dynamic property, and is this even required? It looks like a silent ignore here, as the storagerouter is mandatory for an MDS service.

kvanhijf committed:
Merge branch 'master' into ovs-4348-geoscale

kvanhijf committed:
OVS-4359 - Fixes based on review comments

khenderick commented on ovs/extensions/generic/watcher.py:
```diff
@@ -109,11 +107,10 @@ def services_running(target):
-                    if len(clusters) != 1:
-                        raise ValueError('Expected exactly 1 "{0}" arakoon cluster'.format(ServiceType.ARAKOON_CLUSTER_TYPES.SD))
-                    client = PyrakoonStore(cluster=str(clusters[0].cluster_id))
+                    cluster_name = str(EtcdConfiguration.get('/ovs/framework/arakoon_clusters|voldrv'))
```
Double casting to string

Commit:
Merge branch 'master' into ovs-4348-geoscale

kvanhijf committed:
OVS-4359 - Fixes during testing

kvanhijf committed:
OVS-4359 - Added try - finally statement for locking

kvanhijf committed:
OVS-4359 - Fixed unit tests

kvanhijf committed:
OVS-4455 - Master distributing keys

khenderick commented on ovs/lib/setup.py:
```diff
@@ -258,33 +257,160 @@ def setup_node(node_type=None):
-                node_password = master_password
-                for node_info in SetupController.nodes.itervalues():
+                ip_hostname_map = {cluster_ip: list({node_name, fqdn_name})}
+                for node_host_name, node_info in SetupController.nodes.iteritems():
```
You don't need to loop here

kvanhijf replied:
Actually i do need to loop, because i don't know the hostname of the master node
And i need the hostname to put in the ip_hostname_map
--> will still add a break in if statement

kvanhijf committed:
OVS-4485 - Setting up ETCD proxy alongside ETCD server

kvanhijf committed:
Merge branch 'master' into ovs-4348-geoscale
